core/templates/components/question-directives/question-editor/question-editor.component.spec.ts

```diff
@@ -0,0 +1,395 @@
+
+    ctrl.$onInit();
+
+    ctrl.interactionIsShown = false;
+    ctrl.stateEditorInitialized = false;
+
+    onInteractionIdChangedEmitter.emit();
+    $scope.$apply();
+
+    expect(ctrl.interactionIsShown).toBe(true);
+    expect(ctrl.stateEditorInitialized).toBe(true);
+  });
+
+  it('should get state content save button placeholder', () => {
+    expect(ctrl.getStateContentSaveButtonPlaceholder()).toBe('Save Question');
+  });
+
+  it('should get state content placeholder', () => {
+    expect(ctrl.getStateContentPlaceholder()).toBe('Type your question here.');
+  });
+
+  it('should return undefined when calling functions which are not applicable' +
+    ' to question state', () => {
+    expect(ctrl.navigateToState()).toBe(undefined);
+    expect(ctrl.addState()).toBe(undefined);
+    expect(ctrl.recomputeGraph()).toBe(undefined);
+    expect(ctrl.refreshWarnings()).toBe(undefined);
+  });
+
+  it('should save state content when user clicks on save', () => {
+    expect(ctrl.interactionIsShown).toBe(undefined);
+    expect(ctrl.questionStateData.content)
+      .toEqual(SubtitledHtml.createFromBackendDict({
+        html: 'Question 1',
+        content_id: 'content_1'
+      }));
+
+    ctrl.saveStateContent('New content');
+
+    expect(ctrl.interactionIsShown).toBe(true);
+    expect(ctrl.questionStateData.content).toBe('New content');
+  });
+
+  it('should save interaction ID when interaction is saved', () => {
+    spyOn(StateEditorService, 'setInteractionId');
+
+    ctrl.saveInteractionId('TextInput');
+
+    expect(StateEditorService.setInteractionId).toHaveBeenCalledWith(
+      'TextInput'
+    );
+  });
+
+  it('should save interaction answer groups when interaction is saved', () => {
+    spyOn(StateEditorService, 'setInteractionAnswerGroups');
+
+    ctrl.saveInteractionAnswerGroups('New Answer Group');
+
+    expect(StateEditorService.setInteractionAnswerGroups).toHaveBeenCalledWith(
+      'New Answer Group'
+    );
+  });
+
+  it('should save interaction default outcome when' +
+    ' interaction is saved', () => {
+    spyOn(StateEditorService, 'setInteractionDefaultOutcome');
+
+    ctrl.saveInteractionDefaultOutcome('New outcome');
+
+    expect(StateEditorService.setInteractionDefaultOutcome)
+      .toHaveBeenCalledWith('New outcome');
+  });
+
+  it('should save interaction ID when interaction is saved', () => {
+    spyOn(StateEditorService, 'setInteractionCustomizationArgs');
+
+    ctrl.saveInteractionCustomizationArgs('Customization Args');
+
+    expect(StateEditorService.setInteractionCustomizationArgs)
+      .toHaveBeenCalledWith('Customization Args');
+  });
+
+  it('should set interaction solution when interaction is saved', () => {
+    spyOn(StateEditorService, 'setInteractionSolution');
+
+    ctrl.saveSolution('Solution');
+
+    expect(StateEditorService.setInteractionSolution)
+      .toHaveBeenCalledWith('Solution');
+  });
+
+  it('should save interaction ID when interaction is saved', () => {
```

Comment:
This test desc seems to be repeated. Is this correct?

Reply:
Corrected